com.unity.ml-agents/Runtime/RecursionChecker.cs

```diff
@@ -0,0 +1,35 @@
+using System;
+
+namespace Unity.MLAgents
+{
+    internal class RecursionChecker : IDisposable
+    {
+        private bool m_IsRunning;
+        private string m_MethodName;
+
+        public RecursionChecker(string methodName)
+        {
+            m_MethodName = methodName;
+        }
+
+        public IDisposable Start()
+        {
+            if (m_IsRunning)
+            {
+                throw new UnityAgentsException(
+                    $"{m_MethodName} called recursively. " +
+                    "This might happen if you call EnvironmentStep() or EndEpisode() from custom " +
+                    "code such as CollectObservations() or OnActionReceived()."
```

Comment:
I think it would be more general if instead of passing methodName as argument, we passed the whole error instead. The constructor would look like :
```
new RecursionChecker("CollectObservations() was called recursively. Make sure there are ...");
```
This would make the error messages more clear, and the code future proofer, but I do not have a strong opinion on this.

Reply:
I don't really know how to make the errors messages more clear and still catch all possible permutations that users could set up. I'm afraid if we say something like "don't call EndEpisode from CollectObservations" and someone manages to hit CollectObservations recursively some other way, it will be more confusing.